auth0/src/main/java/com/auth0/android/provider/CustomTabsController.java

```diff
@@ -0,0 +1,190 @@
+    private static final String CHROME_SYSTEM = "com.google.android.apps.chrome";
+    private static final String CHROME_BETA = "com.android.chrome.beta";
+    private static final String CHROME_DEV = "com.android.chrome.dev";
+
+    private final WeakReference<Context> context;
+    private final String preferredPackage;
+    private CustomTabsSession session;
+    private Uri nextUri;
+    private boolean isBound;
+
+    @VisibleForTesting
+    CustomTabsController(@NonNull Context context, @NonNull String browserPackage) {
+        this.context = new WeakReference<>(context);
+        this.preferredPackage = browserPackage;
+    }
+
+    public CustomTabsController(@NonNull Context context) {
+        this(context, getBestBrowserPackage(context));
+    }
+
+    @VisibleForTesting
+    void clearContext() {
+        this.context.clear();
+    }
+
+    @Override
+    public void onCustomTabsServiceConnected(ComponentName componentName, CustomTabsClient customTabsClient) {
+        Log.d(TAG, "CustomTabs Service connected");
+        isBound = true;
+        if (customTabsClient != null) {
+            customTabsClient.warmup(0L);
+            session = customTabsClient.newSession(null);
+        }
+        if (nextUri != null) {
+            launchUri(nextUri);
+        }
+    }
+
+    @Override
+    public void onServiceDisconnected(ComponentName componentName) {
+        Log.d(TAG, "CustomTabs Service disconnected");
+        session = null;
+        isBound = false;
+    }
+
+    /**
+     * Attempts to bind the Custom Tabs Service to the Context.
+     *
+     * @return true if the request to bind the service was successful, false if the service was already bound or it couldn't be bound.
+     */
+    public boolean bindService() {
+        Log.v(TAG, "Trying to bind the service");
+        Context context = this.context.get();
+        boolean success = false;
+        if (!isBound && context != null) {
+            success = CustomTabsClient.bindCustomTabsService(context, preferredPackage, this);
+        }
+        Log.v(TAG, "Bind request result: " + success);
+        return success;
+    }
+
+    /**
+     * Attempts to unbind the Custom Tabs Service from the Context.
+     */
+    public void unbindService() {
+        Log.v(TAG, "Trying to unbind the service");
+        Context context = this.context.get();
+        if (context != null) {
+            context.unbindService(this);
+        }
+        nextUri = null;
+    }
+
+    /**
+     * Attempst to bind the Custom Tabs Service to the Context and opens a Uri as soon as possible.
+     *
+     * @param uri the uri to open in a Custom Tab or Browser.
+     * @return true if the request to bind the service was successful, false if the service was already bound or it couldn't be bound.
+     */
+    public boolean bindServiceAndLaunchUri(@NonNull Uri uri) {
+        nextUri = uri;
+        boolean boundRequestSuccess = bindService();
+        if (isBound || !boundRequestSuccess) {
```

Comment:
ok I get it.
but from the way it's used, this can never be "already bound" because a new controller is created each time. maybe something to optimize in the activity that uses this in https://github.com/auth0/Auth0.Android/pull/95/files#diff-101f8235cd64f6d21cfc3fc82d610bd5R105